File: PlayCover/AppInstaller/optool/patch.m
//
//  patch.m
//  PlayCover
//
//  Created by Александр Дорофеев on 20.11.2021.
//

#import <Foundation/Foundation.h>

#import "patch.h"

BOOL patch_binary_with_dylib(NSString *binaryPath, NSString *dylibName) {
    
    NSLog(@"%@", dylibName);
    // Extract binary data
    NSMutableData *binary = [NSMutableData dataWithContentsOfFile:binaryPath];

    // Extract binary headers
    struct thin_header headers[4];
    uint32_t numHeaders = 0;
    headersFromBinary(headers, binary, &numHeaders);
    
    // Loop through headers
    for (uint32_t i = 0; i < numHeaders; i++) {
        struct thin_header macho = headers[i];
        // Insert dylib load entry into binary
        if (insertLoadEntryIntoBinary(dylibName, binary, macho, LC_LOAD_DYLIB)) {
            LOG("Successfully inserted a command for %s", CPU(macho.header.cputype));
        } else {
            LOG("Failed to insert a command for %s", CPU(macho.header.cputype));
        }
    }
    
    [binary writeToFile:binaryPath atomically:NO];

    return YES;
}

BOOL remove_play_tools_from(NSString *binaryPath, NSString *dylibName) {
    NSMutableData *binary = [NSMutableData dataWithContentsOfFile:binaryPath];
    
    struct thin_header headers[4];
    uint32_t numHeaders = 0;
    headersFromBinary(headers, binary, &numHeaders);
    
    // Loop through headers
    for (uint32_t i = 0; i < numHeaders; i++) {
        struct thin_header macho = headers[i];
        // Insert dylib load entry into binary
        removeLoadEntryFromBinary(binary, macho, dylibName);
    }
    
    [binary writeToFile:binaryPath atomically:NO];
    
    return YES;
}

File: PlayCover/AppInstaller/optool/Patch.swift
//
//  Patch.swift
//  PlayCover
//
//  Created by Isaac Marovitz on 20/09/2022.
//

import Foundation

class Patch {
    @discardableResult
    public static func patchBinaryWithDylib(binaryUrl: URL, dylibName: String) throws -> Bool {
        print(dylibName)

        var binary: OffsetNSData = OffsetNSData(offset: 0,
                                                data: try NSMutableData(contentsOf: binaryUrl))
        var headers: [thin_header] = Array(repeating: thin_header(), count: 4)
        var numHeaders: UInt32 = 0
        Headers.headersFromBinary(&headers, &binary, &numHeaders)

        for macho in headers {
            if Operations.insetLoadEntryIntoBinary(dylibName, &binary, macho) {
                print("Successfully inserted a command for \(cpu(macho.header.cputype))")
            } else {
                print("Failed to insert a command for \(cpu(macho.header.cputype))")
            }
        }

        try binary.data.write(to: binaryUrl)

        return true
    }

    public static func removePlayToolsFrom(binaryUrl: URL, dylibName: String) throws -> Bool {
        var binary: OffsetNSData = OffsetNSData(offset: 0,
                                                data: try NSMutableData(contentsOf: binaryUrl))

        var headers: [thin_header] = Array(repeating: thin_header(), count: 4)
        var numHeaders: UInt32 = 0
        Headers.headersFromBinary(&headers, &binary, &numHeaders)

        for macho in headers {
            Operations.removeLoadEntryFromBinary(&binary, macho, dylibName)
        }

        try binary.data.write(to: binaryUrl)

        return true
    }

    public static func cpu(_ cputype: cpu_type_t) -> String {
        if cputype == CPU_TYPE_I386 {
            return "x86"
        } else if cputype == CPU_TYPE_X86_64 {
            return "x86_64"
        } else if cputype == CPU_TYPE_ARM {
            return "arm"
        } else if cputype == CPU_TYPE_ARM64 {
            return "arm64"
        }
        return ""
    }

    public static func loadCommand(_ loadCommand: UInt32) -> String {
        if loadCommand == LC_REEXPORT_DYLIB {
            return "LC_REEXPORT_DYLIB"
        } else if loadCommand == LC_LOAD_WEAK_DYLIB {
            return "LC_LOAD_WEAK_DYLIB"
        } else if loadCommand == LC_LOAD_UPWARD_DYLIB {
            return "LC_LOAD_UPWARD_DYLIB"
        } else if loadCommand == LC_LOAD_DYLIB {
            return "LC_LOAD_DYLIB"
        }
        return ""
    }
}
